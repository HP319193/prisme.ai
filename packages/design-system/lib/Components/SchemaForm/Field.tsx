import { useEffect, useMemo, useRef, useState } from 'react';
import { Form, useField } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import { SchemaFormContext, useSchemaForm } from './context';
import Enum from './Enum';
import OneOf from './OneOf';
import { FieldProps } from './types';
import { getFieldOptions } from './utils';
import { OnChange } from 'react-final-form-listeners';

export interface FieldComponentProps extends FieldProps {
  components: SchemaFormContext['components'];
}

export const Field = ({ components, ...props }: FieldComponentProps) => {
  const field = useField(props.name, getFieldOptions(props.schema));

  // Update value with default value
  const [_default, setDefault] = useState();
  useEffect(() => {
    if (!_default && props.schema.default) {
      setDefault(props.schema.default);
    }
  }, [_default, props.schema.default]);
  useEffect(() => {
    if (!_default || field.input.value !== undefined) {
      return;
    }
    field.input.onChange(_default);
  }, [_default]);

  const Component = useMemo(() => {
    if (props.schema.oneOf) {
      return OneOf;
    }
    const { 'ui:widget': UiWidget } = props.schema;
    if (typeof UiWidget === 'function') {
      return UiWidget;
    }

    if (UiWidget && components.UiWidgets && components.UiWidgets[UiWidget]) {
      return components.UiWidgets[UiWidget];
    }

    // This widget can be used with any type
    if (UiWidget === 'select') {
      return components.FieldSelect;
    }

    if (props.schema.enum) {
      return Enum;
    }

    switch (props.schema.type) {
      case 'localized:string':
      case 'localized:number':
        return components.FieldLocalizedText;
      case 'string':
      case 'number':
        return components.FieldText;
      case 'boolean':
        return components.FieldBoolean;
      case 'localized:boolean':
        return components.FieldLocalizedBoolean;
      case 'object':
        return components.FieldObject;
      case 'array':
        if (UiWidget === 'tags') {
          return components.FieldTags;
        }
        if (UiWidget === 'upload') {
          return components.FieldArrayUpload;
        }
        return components.FieldArray;
      default:
        return components.FieldAny;
    }
  }, [props.schema]);

  if (props.schema.hidden || !Component) return null;

  return <Component {...props} />;
};

const LinkedField = (props: FieldProps) => {
  const { components } = useSchemaForm();
  const { Field: FieldComponent = Field } = components;
  return <FieldComponent {...props} components={components} />;
};

export const SelfField = (
  props: Pick<FieldProps, 'schema' | 'label'> & {
    value: any;
    onChange: (v: any) => void;
  }
) => {
  const values = useRef({ values: props.value });
  return (
    <Form
      onSubmit={props.onChange}
      initialValues={values.current}
      mutators={{ ...arrayMutators }}
    >
      {() => (
        <>
          <OnChange name="values">
            {(value, previous) => {
              if (previous === value) return;
              props.onChange(value);
            }}
          </OnChange>
          <LinkedField {...props} name="values" />
        </>
      )}
    </Form>
  );
};

export default LinkedField;
