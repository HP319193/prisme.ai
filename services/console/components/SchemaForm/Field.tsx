import { Input, Switch } from '@prisme.ai/design-system';
import { FC } from 'react';
import FieldContainer from '../../layouts/Field';
import { useField } from 'react-final-form';
import { CodeEditorInline } from '../CodeEditor/lazy';
import { FieldValidator } from 'final-form';
import { Schema } from './types';

interface FieldProps {
  field: string;
  type: string;
  required: boolean;
  oneOf?: Schema['oneOf'];
}

export const Field: FC<FieldProps> = ({ field, type, required, oneOf }) => {
  const { input } = useField(field);
  const validate: FieldValidator<any> = (value) => {
    const isRequired = oneOf
      ? oneOf.every(({ required }) => required.includes(field))
      : required;
    return !value && isRequired ? 'required' : undefined;
  };
  switch (type) {
    case 'object':
      return (
        <FieldContainer
          key={field}
          name={field}
          label={field}
          validate={validate}
        >
          {({ input, className }) => (
            <div>
              <CodeEditorInline mode="json" {...input} className={className} />
            </div>
          )}
        </FieldContainer>
      );
    case 'boolean':
      return (
        <div className="mb-5">
          <label className="mx-2 flex flex-1 align-center">
            <Switch
              checked={input.value}
              onChange={(value) => input.onChange(value)}
              className="!mr-2"
            />
            <div>{field}</div>
          </label>
        </div>
      );
    case 'string':
    default:
      return (
        <FieldContainer key={field} name={field} validate={validate}>
          {({ input, className }) => (
            <Input id={field} label={field} {...input} className={className} />
          )}
        </FieldContainer>
      );
  }
};
