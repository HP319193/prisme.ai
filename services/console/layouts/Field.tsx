import { FC, ReactNode } from 'react';
import {
  FieldInputProps,
  FieldMetaState,
  FieldProps as FProps,
  useField,
} from 'react-final-form';
import { isFormFieldValid } from '../utils/forms';

interface FieldProps extends Partial<FProps<any, any>> {
  label?: string | ReactNode;
  className?: string;
  children:
    | ReactNode
    | ((props: {
        className: string;
        input: FieldInputProps<any, HTMLElement>;
        meta: FieldMetaState<any>;
      }) => ReactNode);
}
export const FieldContainer: FC<FieldProps> = ({
  label,
  className,
  children,
  name = '',
  ...fieldProps
}) => {
  const { input, meta } = useField(name, fieldProps);
  return (
    <div className="p-field mb-5">
      <span
        className={`flex ${label ? 'p-float-label mx-2 ' : ''}${
          className || ''
        }`}
      >
        {typeof children === 'function'
          ? children({
              input,
              meta,
              className: `flex flex-1 ${
                isFormFieldValid(meta) ? 'p-invalid' : ''
              }`,
            })
          : children}
        {label && <label htmlFor={name}>{label}</label>}
      </span>
    </div>
  );
};

export default FieldContainer;
