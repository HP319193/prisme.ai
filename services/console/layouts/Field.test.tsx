import Field from "./Field";
import renderer from "react-test-renderer";
import { useField } from "react-final-form";

jest.mock("react-final-form", () => {
  const mock = {
    input: {},
    meta: {},
  };
  return {
    useField: () => mock,
  };
});
it("should render", () => {
  const root = renderer.create(<Field label="foo">Foo</Field>);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should render an input", () => {
  const root = renderer.create(
    <Field label="foo">
      {({ className, input, meta }) => (
        <input className={className} {...input} {...meta} />
      )}
    </Field>
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it("should render an invalid input", () => {
  useField("foo").meta.error = "error";
  useField("foo").meta.touched = true;
  const root = renderer.create(
    <Field label="foo">
      {({ className, input, meta }) => (
        <input className={className} {...input} {...meta} />
      )}
    </Field>
  );
  expect(root.toJSON()).toMatchSnapshot();
});
