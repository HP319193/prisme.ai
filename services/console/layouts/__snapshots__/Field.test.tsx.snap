// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="p-field mb-5"
>
  <span
    className="flex p-float-label mx-2 undefined"
  >
    Foo
    <label
      htmlFor=""
    >
      foo
    </label>
  </span>
</div>
`;

exports[`should render an input 1`] = `
<div
  className="p-field mb-5"
>
  <span
    className="flex p-float-label mx-2 undefined"
  >
    <input
      className="flex flex-1 "
    />
    <label
      htmlFor=""
    >
      foo
    </label>
  </span>
</div>
`;

exports[`should render an invalid input 1`] = `
<div
  className="p-field mb-5"
>
  <span
    className="flex p-float-label mx-2 undefined"
  >
    <input
      className="flex flex-1 p-invalid"
      error="error"
      touched={true}
    />
    <label
      htmlFor=""
    >
      foo
    </label>
  </span>
</div>
`;
